File: src/main/java/wood/mike/client/SeClient.java
package wood.mike.client;

import io.grpc.ManagedChannel;
import io.grpc.ManagedChannelBuilder;
import io.grpc.stub.StreamObserver;
import wood.mike.*;

import java.util.Iterator;
import java.util.Scanner;
import java.util.concurrent.CountDownLatch;
import java.util.concurrent.TimeUnit;
import java.util.stream.Stream;

public class SeClient {

    private static final String HOST = "localhost";
    private static final int PORT = 50052;

    private final SecretEscapesGrpc.SecretEscapesBlockingStub blockingStub;
    private final SecretEscapesGrpc.SecretEscapesStub nonBlockingStub;
    private final ManagedChannel channel;
    private final Scanner scanner;

    public SeClient() {
        channel = ManagedChannelBuilder.forAddress(HOST, PORT)
                .usePlaintext()
                .build();

        blockingStub = SecretEscapesGrpc.newBlockingStub(channel);
        nonBlockingStub = SecretEscapesGrpc.newStub(channel);
        scanner = new Scanner(System.in);
    }

    private void run() throws InterruptedException {
        String result = switch (getChoice()) {
            case 1 -> unaryExample();
            case 2 -> serverStreamingExample();
            case 3 -> clientStreamingExample();
            case 4 -> bidirectionalStreamingExample();
            default -> throw new IllegalStateException("Unexpected value");
        };

        System.out.println(result);
        close();
    }

    private String unaryExample() {
        var seResponse =
                blockingStub.getSeSale(
                        SeSaleRequest
                        .newBuilder()
                        .setId(1).build());
        return String.format("Server reply: %s", seResponse);
    }

    private String serverStreamingExample() {
        Iterator<SeSaleReply> seSaleReplyIterator =
                blockingStub.getAllSeSales(SeSaleRangeRequest
                        .newBuilder()
                        .setStart("2025-02-01")
                        .setEnd("2025-02-07")
                        .build());

        seSaleReplyIterator.forEachRemaining(System.out::println);
        return "Done";
    }

    private String clientStreamingExample() throws InterruptedException {
        CountDownLatch latch = new CountDownLatch(1);

        StreamObserver<BatchSeSaleReply> responseObserver = new StreamObserver<>() {
            @Override
            public void onNext(BatchSeSaleReply batchSeSaleReply) {
                batchSeSaleReply.getRepliesList().forEach(req ->
                        System.out.println("SE sale: " + req)
                );
            }

            @Override
            public void onError(Throwable t) {
                System.err.println("Error from server: " + t.getMessage());
                latch.countDown();
            }

            @Override
            public void onCompleted() {
                System.out.println("Server finished processing.");
                latch.countDown();
            }
        };

        StreamObserver<SeSaleRequest> requestObserver =
                nonBlockingStub.getBatchedSeSales(responseObserver);

        Stream.of(1,2,3,4,5).map(id ->
            SeSaleRequest.newBuilder()
                .setId(id)
                .build()
        ).forEach(requestObserver::onNext);

        requestObserver.onCompleted();

        latch.await(5, TimeUnit.SECONDS);

        return "Done";
    }

    private String bidirectionalStreamingExample() throws InterruptedException {
        // Latch to wait for the server responses before shutting down
        CountDownLatch latch = new CountDownLatch(1);

        // Handle server responses
        StreamObserver<PackageRequest> requestObserver =
                nonBlockingStub.customizePackage(new StreamObserver<PackageResponse>() {
            @Override
            public void onNext(PackageResponse response) {
                System.out.println("Received response:");
                System.out.println("Type: " + response.getSelectionType());
                System.out.println("ID: " + response.getSelectionId());
                System.out.println("Available: " + response.getAvailable());
                System.out.println("Price: $" + response.getPrice());
                System.out.println("Message: " + response.getMessage());
                System.out.println("----------------------------------");
            }

            @Override
            public void onError(Throwable t) {
                System.err.println("Error from server: " + t.getMessage());
                latch.countDown();
            }

            @Override
            public void onCompleted() {
                System.out.println("Server completed sending responses.");
                latch.countDown();
            }
        });

        PackageRequest flightRequest = PackageRequest.newBuilder()
                .setUserId("user123")
                .setSelectionType("flight")
                .setSelectionId("FL123")
                .setQuantity(1)
                .build();

        PackageRequest hotelRequest = PackageRequest.newBuilder()
                .setUserId("user123")
                .setSelectionType("hotel")
                .setSelectionId("HT456")
                .setQuantity(3)
                .build();

        PackageRequest tourRequest = PackageRequest.newBuilder()
                .setUserId("user123")
                .setSelectionType("tour")
                .setSelectionId("TR789")
                .setQuantity(2)
                .build();

        System.out.println("Sending package customization requests...");
        requestObserver.onNext(flightRequest);
        requestObserver.onNext(hotelRequest);
        requestObserver.onNext(tourRequest);

        // Mark the end of requests
        requestObserver.onCompleted();

        // Wait until server finishes sending responses
        latch.await(5, TimeUnit.SECONDS);

        return "Done";
    }

    private void close() {
        scanner.close();
        channel.shutdown();
    }


    private int getChoice() {
        System.out.println(
                """
                        1. Fetch one SE sale - Unary
                        2. Fetch multiple SE sales - server streaming
                        3. Fetch multiple SE sales batched - client streaming
                        4. Build package - bidirectional streaming""");
        return scanner.nextInt();
    }

    public static void main(String[] args) throws InterruptedException {
        new SeClient().run();
    }
}

File: python_src/se_client.py
from time import sleep
import grpc
import SE_pb2
import SE_pb2_grpc

GRPC_PORT = 50052



def run():
    with grpc.insecure_channel(f"localhost:{GRPC_PORT}") as channel:
        stub = SE_pb2_grpc.SecretEscapesStub(channel)
        print("1. Fetch one SE sale - Unary")
        print("2. Fetch multiple SE sales - server streaming")
        print("3. Fetch multiple SE sales batched - client streaming")
        print("4. Build package - bidirectional streaming")

        rpc_call = input("Which rpc call do you wish to run? ")

        if rpc_call == "1":
            unary_example(stub)
        elif rpc_call == "2":
            server_streaming_example(stub)
        elif rpc_call == "3":
            client_streaming_example(stub)
        elif rpc_call == "4":
            bidirectional_example(stub)

def unary_example(stub):
    se_sale_request = SE_pb2.SeSaleRequest(id = 1)
    se_sale_reply = stub.GetSeSale(se_sale_request)
    print('SeSaleRequest Response received: "%s"' % se_sale_reply)

def server_streaming_example(stub):
    se_sale_range_request = SE_pb2.SeSaleRangeRequest(start = "2025-01-19", end = "2025-01-26")
    se_replies = stub.GetAllSeSales(se_sale_range_request)
    for se_reply in se_replies:
        print('SeSaleRangeRequest Response received: "%s"' % se_reply)

def client_streaming_example(stub):
    def generate_requests():
        sale_ids = [1, 2, 3, 4, 5]
        for sale_id in sale_ids:
            request = SE_pb2.SeSaleRequest(id=sale_id)
            print(f"Sending sale request for ID: {sale_id}")
            yield request
            sleep(0.2)  # Simulate slight delay between requests

    response = stub.GetBatchedSeSales(generate_requests())

    print("\nReceived batched sales reply:")
    for sale_reply in response.replies:
        print(f"Sale ID: {sale_reply.id}, URL: {sale_reply.url_slug}")

def bidirectional_example(stub):
    def generate_requests():
        requests = [
            SE_pb2.PackageRequest(user_id="user123", selection_type="flight", selection_id="FL123", quantity=1),
            SE_pb2.PackageRequest(user_id="user123", selection_type="hotel", selection_id="HT456", quantity=2),
            SE_pb2.PackageRequest(user_id="user123", selection_type="tour", selection_id="TR789", quantity=1),
        ]

        for req in requests:
            print(f"Sending request: {req.selection_type} - {req.selection_id}")
            yield req
            sleep(0.5)  # Simulate delay between requests

    responses = stub.CustomizePackage(generate_requests())

    for response in responses:
        print(f"Received response: {response.selection_type} - {response.selection_id}")
        print(f"Available: {response.available}, Price: ${response.price}, Message: {response.message}")
        print("-" * 40)

if __name__ == '__main__':
    run()